### PR TITLE
Support object-space normal maps

### PR DESCRIPTION
**Pull request by WestLangley**
As proposed in #14139.

### CODE REVIEW

**Comment by WestLangley** [comment]
/ping @Julien-Arlandis
/ping @Ben-Mack

Usage:
```javascript
material.normalMapType = THREE.ObjectSpaceNormalMap; // default: THREE.TangentSpaceNormalMap
```

@bhouston nomenclature can be changed later.

An example can also be added later if desired. This PR is backwards-compatible.

**Review by Julien-Arlandis — APPROVED** [review]

**Comment by pailhead** [comment]
Code illustration for this problem:

`my_normal_fragment_maps.glsl`
```glsl 
normal = texture2D( normalMap, vUv ).xyz * 2.0 - 1.0; 

#ifdef FLIP_SIDED
  normal = - normal;
#endif

#ifdef DOUBLE_SIDED
  normal = normal * ( float( gl_FrontFacing ) * 2.0 - 1.0 );
#endif

normal = normalize( normalMatrix * normal );
```


```javascript

var myMaterial = new Material()

myMaterial.onBeforeCompile = shader =>{ shader.fragmentShader = shader.fragmentShader.replace('#include <normal_fragment_maps>, require('my_normal_fragment_maps.glsl') }
```

I read this as one line of JS, and some data stored as a string. This PR touches 13 files, for what may end up being dead code for most people. Say if all the apps in existence got updated to the version of three.js that has this, it would be dead code.

In comparison, https://github.com/mrdoob/three.js/pull/14231 is also hidden behind optional flags, meaning it's dead code if user's don't opt in but it's being rejected over and over.

**Comment by pailhead** [comment]
In comparison, a core change to three.js that should allow this to not be in core, is roughly 20 lines of calls:
https://github.com/mrdoob/three.js/pull/14231

To reiterate, it can be used to make this PR optional. 

This touches 60, and introduces something that no one has ever used with three.js so far (every existing app out there, doesn't need this code).

**Comment by bhouston** [comment]
This is a great change @WestLangley.   Thank you.

**Review by Mugen87 — APPROVED** [review]

**Comment by Julien-Arlandis** [comment]
Very good work/ Thanks !

**Comment by Julien-Arlandis** [comment]
3Dmap with the new normal map configuration : http://vm1.3dmap.fr:1357/mapRender/?ref=savoie
How to know when the commit will be passed on the official lib?

**Comment by WestLangley** [comment]
>How to know when the commit will be passed on the official lib?

It may not have been merged because @pailhead has requested it not be.

**Comment by pailhead** [comment]
>It may not have been merged because @pailhead has requested it not be.

Eh, when you put it like this.... :smile: 

I don't want to be the one blocking this, but i do want to start a discussion that may block this (in the current context).

Issues:

1. extra code
2. lots of files touched
3. should it be on the material, or on the map? 
```javascript
var myNormalMap = new Texture({normal: THREE.ObjectSpaceNormal})
```
4. is the best way to do this with defines, could it be different chunks for legibility? (don't overload `normalmap_pars_fragment` and `normal_fragment_maps` have the Renderer use the appropriate chunk based on the type of normal map)


On 1 and 2, i'm curious how is this request different from any other request that's asking for something to be done that's not in the examples. On 1 in particular, it's not that much code, but it is extra code. 2 complicates the core, which may not be the best thing to do given the `NodeMaterial` proposal (these i imagine would be two different normal mapping nodes, hence it ties into 4).

On 3 and 4  it also ties into this user request, just because someone requested it shouldn't mean that we should just scramble to land this ASAP, i think it needs to be reviewed.

**Comment by pailhead** [comment]
In the meantime, @Julien-Arlandis you can modify the shader to do this. You'd have to wait for this to be released with the next version anyway, so the modification can be made available much sooner. I'll post a working example.

**Comment by pailhead** [comment]
Related:

The three.js community has started looking into other engines and tools for integration and influence. I'm not sure if three will be merged with Sea3D, but Unity too was cited as influence for https://github.com/mrdoob/three.js/issues/7522.

If we are looking at unity for guidance, this is what a texture object looks like in unity:
unity normal maps
map_type

The material then just takes the texture. I don't just putting all this stuff on the `Material` and coupling it with `WebGLRenderer` in so many places. I'd prefer something more generic going on, where the renderer can set this up properly in the most abstract way possible. In the future it could figure out which node to use or switch out, now, it could be a different Chunk.

**Comment by bhouston** [comment]
Why is @pailhead capable of blocking perfectly good PRs?

**Comment by pailhead** [comment]
@pailhead is not capable of blocking perfectly good PRs. This being open source and all, open to public and such, I can question whether this is a perfectly good PR.

**Comment by pailhead** [comment]
If i know that the texture i'm loading is a normal map, and that it's in object space, i don't want to have to inform every material in the scene (and out of the scene) that it needs to treat that map as such.

```javascript
var my_shared_OBJECT_SPACE_normal_map = loadTexture()

scene.traverse( o=>{
if(o.material.normalMap === my_shared_OBJECT_SPACE_normal_map){
  o.material.objectSpaceNormalMap = true
}
})
```

I'd rather have:

```javascript
var my_shared_OBJECT_SPACE_normal_map = loadTexture()
my_shared_OBJECT_SPACE_normal_map.space = "object"
```

And the rest takes care of itself ^

**Comment by bhouston** [comment]
For example, I didn't like how animations were done in Three.JS.  Instead of blocking other people's contributions to ThreeJS I just rewrote the animation system into the form I thought it should be in: https://github.com/mrdoob/three.js/pull/6934

When I didn't like the fact that textures didn't support encodings, I didn't just complain about it, I wrote a solution in the form I thought was best: https://github.com/mrdoob/three.js/pull/8117/files

Be the change you want to see in the world.  Otherwise you will not be happy and you will feel as if you are at the mercy of others.

**Comment by pailhead** [comment]
@bhouston 

Actually tl:dr;

>Be the change you want to see in the world. Otherwise you will not be happy and you will feel as if you are at the mercy of others.

I want to do exactly this! But i can't shake of the feeling of mercy of others. If i want to do these PR's i reach out to tools that i don't like but my PRs to fix them have been rejected. An argument that i should use my proposal as my own tool for developing three.js would actually probably be pretty valid, but short of a `wait for NodeMaterial` generic reply i get nothing, so it shakes up my confidence and hurts my feelings :)

File: src/renderers/webgl/WebGLProgram.js [hunk 1]
```diff
@@ -455,6 +456,7 @@ function WebGLProgram( renderer, extensions, code, material, shader, parameters
 			parameters.emissiveMap ? '#define USE_EMISSIVEMAP' : '',
 			parameters.bumpMap ? '#define USE_BUMPMAP' : '',
 			parameters.normalMap ? '#define USE_NORMALMAP' : '',
+			( parameters.normalMap && parameters.objectSpaceNormalMap ) ? '#define OBJECTSPACE_NORMALMAP' : '',
```
Comment (pailhead) [inline src/renderers/webgl/WebGLProgram.js]: If this got changed to:
```javascript
( parameters.normalMap && parameters.normalMap.objectSpaceNormalMap ) ? '#define OBJECTSPACE_NORMALMAP' : '',
```
I, three.js user pailhead, will be happy.

File: src/renderers/webgl/WebGLPrograms.js [hunk 2]
```diff
@@ -148,6 +148,7 @@ function WebGLPrograms( renderer, extensions, capabilities ) {
 			emissiveMapEncoding: getTextureEncodingFromMap( material.emissiveMap, renderer.gammaInput ),
 			bumpMap: !! material.bumpMap,
 			normalMap: !! material.normalMap,
+			objectSpaceNormalMap: material.normalMapType === ObjectSpaceNormalMap,
```
Comment (pailhead) [inline src/renderers/webgl/WebGLPrograms.js]: Here:
```
objectSpaceNormalMap: material.normalMap.objectSpace,
```

File: src/materials/MeshPhongMaterial.js [hunk 3]
```diff
@@ -136,6 +138,7 @@ MeshPhongMaterial.prototype.copy = function ( source ) {
 	this.bumpScale = source.bumpScale;
 
 	this.normalMap = source.normalMap;
+	this.normalMapType = source.normalMapType;
```
Comment (pailhead) [inline src/materials/MeshPhongMaterial.js]: nuke

File: src/materials/MeshPhongMaterial.js [hunk 4]
```diff
@@ -79,6 +80,7 @@ function MeshPhongMaterial( parameters ) {
 	this.bumpScale = 1;
 
 	this.normalMap = null;
+	this.normalMapType = TangentSpaceNormalMap;
```
Comment (pailhead) [inline src/materials/MeshPhongMaterial.js]: nuke

File: src/materials/MeshNormalMaterial.js [hunk 5]
```diff
@@ -70,6 +73,7 @@ MeshNormalMaterial.prototype.copy = function ( source ) {
 	this.bumpScale = source.bumpScale;
 
 	this.normalMap = source.normalMap;
+	this.normalMapType = source.normalMapType;
```
Comment (pailhead) [inline src/materials/MeshNormalMaterial.js]: nuke

File: src/materials/Material.js [hunk 6]
```diff
@@ -194,6 +194,7 @@ Material.prototype = Object.assign( Object.create( EventDispatcher.prototype ),
 		if ( this.normalMap && this.normalMap.isTexture ) {
 
 			data.normalMap = this.normalMap.toJSON( meta ).uuid;
+			if ( this.normalMapType !== TangentSpaceNormalMap ) data.normalMapType = this.normalMapType;
```
Comment (pailhead) [inline src/materials/Material.js]: nuke, would be serialized with the texture, there are possibly materials that extend `Material` that do not use normal maps, like maybe depth. Consider not making the super class aware of the sub classes.

**Review by pailhead — COMMENTED** [review]
consider moving this property to `Texture`

File: docs/api/materials/MeshPhongMaterial.html [hunk 7]
```diff
@@ -178,6 +178,13 @@ <h3>[property:Texture normalMap]</h3>
 			the way the color is lit. Normal maps do not change the actual shape of the surface, only the lighting.
 		</p>
 
+		<h3>[property:Integer normalMapType]</h3>
+		<p>
```
Comment (pailhead) [inline docs/api/materials/MeshPhongMaterial.html]: Consider consolidating multiple document entries by moving the property to `Texture`. If not, add the documentation for `MeshNormalMaterial`.

File: docs/api/materials/MeshStandardMaterial.html [hunk 8]
```diff
@@ -220,6 +220,13 @@ <h3>[property:Texture normalMap]</h3>
 			the way the color is lit. Normal maps do not change the actual shape of the surface, only the lighting.
 		</p>
 
+		<h3>[property:Integer normalMapType]</h3>
```
Comment (pailhead) [inline docs/api/materials/MeshStandardMaterial.html]: Consider consolidating multiple document entries by moving the property to `Texture`. If not, add the documentation for `MeshNormalMaterial`.

**Review by pailhead — COMMENTED** [review]
Consider completing or changing the documentation.

**Comment by pailhead** [comment]
Review submitted as feedback, no requests for change. The only thing i would urge you to do, in case you think the feedback is invalid, is to add the missing documentation for `MeshNormalMaterial` because as is only `MeshStandardMaterial` and `MeshPhongMaterial` are documented.

**Comment by bhouston** [comment]
@pailhead  I find you ignored completely the main advice I offered: to allow this to be merged (don't act as a blocker) and then suggest your refactoring in a separate PR (be the chance you want to see.)  I am not sure there is much more to talk about, I have other things to do.

**Comment by WestLangley** [comment]
>add the missing documentation for MeshNormalMaterial because as is only MeshStandardMaterial and MeshPhongMaterial are documented.

The `MeshNormalMaterial` documentation is significantly deficient as-is, and since the deficiencies are unrelated to the feature added here, I decided that document can best be updated in a separate PR.

**Comment by pailhead** [comment]
@WestLangley 

Would you consider not overloading materials, but rather using this as a property of `Texture`?

**Comment by WestLangley** [comment]
@pailhead Coupling a modification of the `Texture` class with this PR is probably not advisable. If you want to get significant changes merged, it's wise to get buy-in from @mrdoob first -- before you invest a lot of effort in the implementation details.

If you feel you have a better design for the `Texture` class, outlining your proposal in a separate issue would be advisable. It would be high-risk, I expect, so be prepared; you may have more success working on something that does not impact significantly on the three.js design.

**Comment by looeee** [comment]
@pailhead if you want to do something really useful, the documentation always needs work. PRs for docs are much easier to get accepted as well.

**Comment by pailhead** [comment]
@looeee 

I spent 20 minutes refactoring this PR, i think that was a decent use of my time. In the process i believe it addressed the documentation, if you compare, the same entry appears on two pages here, only once in the other PR. 

@mrdoob once advised that it's better to push PRs with less code, https://github.com/mrdoob/three.js/pull/14291 is roughly 30% smaller.

**Comment by pailhead** [comment]
I think #14291 is also a good example of the benefits of collaborative effort and peer reviews. 

Objectively, i think it proves that by working together we can shave off that 30% of code.

Subjectively i think it makes it a bit more clean and readable. `Material` is not polluted with random properties from random sub classes (of which there could be many) and the attribute giving a hint what the pixels in `Texture.image` contain live on the `Texture. 

Objectively, i've seen no screen shots, together, we have a view into how @WestLangleys shader code works. 

objectspacenormalmap

**Comment by pailhead** [comment]
Fearing an imminent ban, i'll still go ahead and post this as i think it's useful:

I baked a simple sphere with 3d studio max

http://dusanbosnjak.com/test/webGL/three-normalmap-space/

The example uses the current build and `onBeforeCompile`, but there's a comment which code to remove and which code to enable to make it work with this PR, if anyone is interested. I guess this is the last of me, thank you very much for making an awesome library! Good bye :wave:

**Comment by mrdoob** [comment]
Sorry for the delay merging this one guys. I was out last week on a conference and seems like things got a little bit out of control. 

@pailhead I prefer having this as a property of `Material` rather than a property of `Texture`.

**Comment by mrdoob** [comment]
Thanks @WestLangley!

**Comment by mrdoob** [comment]
The model from #14139 looks great!

screen shot 2018-06-17 at 6 37 28 pm

**Pull request merged**